doc/callbackinfo.md

```diff
@@ -0,0 +1,99 @@
+**WORK IN PROGRESS, NOT YET COMPLETE**
+
+# Callbackinfo
+
+The object representing the components of the JavaScript request being made.
+
+The Callbackinfo object is usually created and passed by the calling system.
+
+The Callbackinfo object contains the arguments passed by the caller. The number of arguments is returned by the `Length` method. Each individual argument can be accessed using the `operator[]` method.
+
+The `SetData` and `Data` methods are used to set and retrieve the data pointer contained in the Callbackinfo object.
+
+## Methods
+
+### Constructor
+
+```cpp
+CallbackInfo(napi_env env, napi_callback_info info);
+```
+
+- `[in] env`: The `napi_env` environment in which to construct the `CallbackInfo` object.
+- `[in] info`: The `napi_callback_info` data structure from which to construct the `CallbackInfo` object.
+
+### Env
+
+```cpp
+Napi::Env Env() const;
+```
+
+Returns the `Env` object in which the request is being made.
+
+### NewTarget
+
+```cpp
+Value NewTarget() const;
+```
+
+Returns the `new.target` value of the constructor call. If the CallbackInfo is not a constructor call, a call to `IsEmpty()` on the returned value returns true.
+
+### IsConstructCall
+
+```cpp
+bool IsConstructCall() const;
+```
+
+Returns a `bool` indicating if the CallbackInfo is a constructor call.
+
+### Length
+
+```cpp
+size_t Length() const;
+```
+
+Returns the number of arguemnts passed in the Callabckinfo object.
+
+### operator []
+
+```cpp
+const Value operator [](size_t index) const;
+```
+
+- `[in] index`: The zero-based index of the requested argument.
+
+Returns a `Value` object containing the requested argument.
+
+### This
+
+```cpp
+Value This() const;
+```
+
+Returns the JavaScript `this` value for the call
+
+### Data
```

Comment:
Would be good to provide a bit more details on what/how Data is used but that can be added in a follow on PR.

Reply:
I'm not clear how Data is used. I'm assuming that it's a place the C++ programmer can store general information with the C++ object, but I'm not sure. If so, I can add that to the documentation.

Reply:
I think we can add the info as a follow on PR we just need to record somewhere that we should come back and do that.  We should generate a test that uses it which would provide an example to base the doc on.